tests/sentry/utils/test_pipeline.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import absolute_import
+
+from django.http import HttpRequest
+
+from sentry.testutils import TestCase
+from sentry.utils.pipeline import (
+    PipelineableProvider,
+    PipelineView,
+    ProviderPipeline,
+)
+
+
+class PipelineStep(PipelineView):
+    def dispatch(self, request, pipeline):
+        pipeline.dispatch_count += 1
+        pipeline.bind_state('some_state', 'value')
+
+
+class DummyProvider(PipelineableProvider):
+    key = 'dummy'
+    pipeline = [PipelineStep(), PipelineStep()]
+
+    def get_pipeline(self):
+        return self.pipeline
+
+
+class DummyProviderManager(object):
+    def get(self, provider):
+        return DummyProvider()
```

Comment:
I don't really have the context for how this will be used in practice so forgive me if this is off the mark, but I'm not sure why we need a `ProviderManager` to fetch a `PipelineableProvider` to fetch a list of `PipelineView`s. Could we dispense with 2 of these levels of abstraction and just initialize our `ProviderPipeline` with the list of views directly? Or are there more complex use cases in the works that need multiple types of providers and provider managers?

Reply:
> Just for completeness of testing, can this return a provider from a keyed list

love it fe31613

> but I'm not sure why we need a `ProviderManager` to fetch a `PipelineableProvider` to fetch a list of `PipelineView`s. Could we dispense with 2 of these levels of abstraction and just initialize our `ProviderPipeline` with the list of views directly? Or are there more complex use cases in the works that need multiple types of providers and provider managers?

Yes, the primary use for the pipeline utility class is to handle OAuth (and SAML) auth flows, as well as any additional setup screens (for example, with github SSO setup you have to select the GH organization you want to do SSO with).

So in that case, the manager is used to lookup what provider pipeline should be used (for example, if logging in via SSO it would use the a provider keyed on 'github', 'google', etc). Then some of the views themselves are sometimes shared (for example, the OAuth pipeline provider shares a view for callback handling)

There are a few different pipelines (Auth: For SSO login authentication and SSO auth setup, Identity: For handling the actual identity lookup through identity providers like google, and Integration: for setting up an integration (which similar to auth, composes identity pipelines).

This should make more sense with the upcoming identity and integration stuff I'm finishing up.